main.rb

```diff
@@ -0,0 +1,83 @@
+module Enumerable
+  def my_each(&prc)
+    i = 0
+    while i < length
+      if is_a?(Array)
+        yield self[i]
+      elsif prc.arity == 1
+        yield(assoc keys[i])
+      else
+        yield(keys[i], self[keys[i]])
+      end
+      i += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    idx = 0
+    while idx < length
+      yield(self[idx], idx)
+      idx += 1
+    end
+    self
+  end
+
+  def my_select
+    if is_a?(Array)
+      selected = []
+      my_each { |ele| selected << ele if yield ele }
+    else
+      selected = {}
+      my_each { |key, value| selected[key] = value if yield(key, value) }
+    end
+    selected
+  end
+
+  def my_all?(&prc)
+    my_each do |ele|
+      return false if prc.call(ele) == false
+    end
+    true
+  end
+
+  def my_any?(&prc)
+    my_each do |ele|
+      return true if prc.call(ele) == true
+    end
+    false
+  end
+
+  def my_none?(&prc)
+    my_each do |ele|
+      return false if prc.call(ele) == true
+    end
+    true
+  end
+
+  def my_count
+    count = 0
+    my_each do |_|
+      count += 1
+    end
+    count
+  end
+
+  def my_map(&prc)
+    mapped = []
+    my_each { |ele| mapped << prc.call(ele) }
+    mapped
+  end
+
+  def my_inject
+    acc = shift # [1,2,3,4,5]
+    item_removed = acc # 1
+    my_each { |num| acc = yield(acc, num) }
+    unshift(item_removed)
+    acc
+  end
```

Comment:
`my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```